Feature/demo fleet adapter

xiyuoh opened the pull request:
## New feature implementation

### Implemented feature

Currently RMF communicates with the full control fleet adapter which sends/receives `rmf_fleet_msgs` to/from slotcar. This PR sets up a mock fleet manager/adapter to replace full control for selected demo fleets using the [fleet_adapter_template](https://github.com/open-rmf/fleet_adapter_template).

### Implementation description

The previous full control fleet adapter is now split into two parts: fleet manager and fleet adapter. The fleet adapter communicates with RMF, then sends navigation waypoints to, or receives robot state information from, the fleet manager server via REST API (FastAPI). At the same time, the fleet manager subscribes to slotcar's `RobotState` and publishes `PathRequest` with `rmf_fleet_msgs`.

This is implemented on TinyRobot, DeliveryRobot, CleanerBotA and CleanerBotE which are full control fleets. Instead of having the fleet's parameters listed in their adapter launch files, they are now consolidated in a `config.yaml` stored in `rmf_demos/config/<demo_name>/`. The robot configurations are also stored here, such as robot name, map name, starting waypoint and charger waypoint. Multiple robots are supported. 

Steps to launch the demos remain the same. Additionally, you can make use of FastAPI's automatic documentation by visiting the fleet's `base_url` + `/docs` to view the fleet manager endpoints and interact with them.

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_manager.py:
```diff
@@ -0,0 +1,259 @@
+            RobotState,
+            'robot_state',
+            self.robot_state_cb,
+            10)
```
```suggestion
            100)
```

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+                self.on_waypoint = None
+                self.on_lane = None
+            time.sleep(0.5)
+            # ------------------------ #
```
So far we have been using docking as a way to initiate cleaning processes. But it doesn't look like that is supported with this fleet adapter. For example, in the `airport_terminal` world, if I request for `zone_1` to be cleaned, once the robot arrives as `zone_1` waypoint, the fleet adapter will call the `dock()` functions. But in the implementation here, it makes a call to `api.start_process()` but the `/open-rmf/rmf_demos_fm/start_task?robot_name={self.robot_name}` endpoint does not trigger anything in `fleet_manager.py` so the robot remains stuck. See related comment in `fleet_manager.py`.

Yadunund replied:
We should also subscribe to `/dock_summary` here to get the list of waypoints the robot will navigate through while docking/cleaning. The in the "Implement your code here" block, we should get the [unstable participant](https://github.com/open-rmf/rmf_ros2/blob/46c2bb931342673347adaf0c5b91f2f4ef9441d4/rmf_fleet_adapter_python/src/adapter.cpp#L137-L143) and then [set the itinerary](https://github.com/open-rmf/rmf_ros2/blob/46c2bb931342673347adaf0c5b91f2f4ef9441d4/rmf_fleet_adapter/src/full_control/main.cpp#L439-L443) with the waypoints remaining in the docking/cleaning path.

xiyuoh replied:
Sorry didn't realise the robot was stuck! Thanks for pointing it out. Added the subscribers in both `fleet_manager` and `RobotCommandHandle`.\
One thing though, `mock_docker` publishes `/dock_summary` before the two subscribers are ready, so the message is often missed. I thought the `qos_profile` specified for this publisher should help with this, but it didn't seem to work for me. I added a `time.sleep(20)` in `mock_docker` before it publishes as a temporary solution, but I'm still looking for a better way in case the manager or command handle takes too long to set up in other cases

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+                            else:
+                                if self.path_index is not None:
+                                    d = (waypoint_wait_time - time_now).seconds
+                                    self.node.get_logger().info(
```
Lets convert this to `self.node.get_logger().debug(...)`

xiyuoh replied:
Got it

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+                    self.state == RobotState.WAITING):
+                # Check if we need to abort
+                if self._quit_path_event.is_set():
+                    self.node.get_logger().info("Aborting previously followed "
```
Include `robot_name` in the printout

xiyuoh replied:
Done

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+
+        return (first, second)
+
+    # def get_remaining_waypoints(self, waypoints: list):
```
Delete commented code

xiyuoh replied:
Done

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotClientAPI.py:
```diff
@@ -0,0 +1,190 @@
+    def process_completed(self):
+        ''' Return True if the robot has successfully completed its previous
+            process request. Else False.'''
+        response = self.data()
```
Instead of duplicating the code here, lets just call `return self.navigation_completed()`

xiyuoh replied:
Done

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+            # IMPLEMENT YOUR CODE HERE #
+            # With whatever logic you need for docking #
+            # ------------------------ #
+            while (not self.api.docking_completed()):
```
I think this is supposed to be `while (not self.api.process_completed()):`

xiyuoh replied:
Yup you're right! Edited

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_manager.py:
```diff
@@ -0,0 +1,259 @@
+            if (robot_name not in self.robots or
+                    len(task.task) < 1):
+                return data
+            # ------------------------ #
```
Implementing this endpoint is crucial for cleaning and docking related tasks. Lets have this node subscribe to `/dock_summary` and store a dict of {dock_name: waypoints} as populated in the `DockSummary` msg. Apart from `robot_name`, we should also receive the `process_name`. This would be the `dock_name` in `RobotCommandHandle::dock(dock_name:str)` . So when we get a request at this endpoint, we lookup the corresponding waypoints for the `dock_name` and send a single `PathRequest` to `slotcar` with the `path` field populated with all the waypoints.

xiyuoh replied:
Ah my bad, I was confused about the cleaning path, didn't realise it was stuck. Made the necessary changes, `fleet_manager` now sends out a `PathRequest` populated with the docking waypoints

Yadunund reviewed (changes requested):
Thanks for all the great work so far! I've left several comments that need to be addressed. Also, in my testing, I'm still encountering the problem where the fleet adapter will crash due to assert failure as it is unable to get the position of the robot from the fleet_manager. Since every simulation world will have a different startup time, it's hard to generalize how many "retries" to perform in `RobotCommandHandle`. I think we should re-architect the adapter as follows to address this:

1. Remove `robot_name` from the constructor of `RobotClientAPI`. Instead each function in the class will now accept the `robot_name` as a parameter. Then we create a single instance of `RobotClientAPI` in `fleet_adapter.py::initialize_fleet()` which will then be shared among all the `RobotCommandHandles` (The constructor should be modified to accept the REST client directly). This would also save the overhead of having so many clients.
2. Instead of this [blob](https://github.com/xiyuoh/rmf_demos/blob/378640287a1054d734469d8d148d3accfc1b0b7b/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_adapter.py#L163-L198) which directly adds robots to the fleet, let's start a separate thread that basically queries `api.data(robot_name)` for each robot name in the config. Only when this data is available, do we create the command_handle and subsequently call `fleet.add_robot()`. We should compute the `Planner::Start` for the robot and pass it as an argument to the constructor of the command_handle. If we are unable to compute a valid start, then we do not add_robot(). We keep track of which robots have not been added and terminate the thread when all of them are added. 
Let me know if you have any questions.

Yadunund commented:
Also, it would be great if we could add a Readme inside `rmf_demos_fleet_adapter` that documents the API endpoitns and the response/request schemas.

xiyuoh commented on rmf_demos/launch/include/adapters/cleanerBotA_adapter.launch.xml:
```diff
@@ -19,61 +17,11 @@
-  
-      <arg name="recharge_threshold" value="0.2"/>
+      <!-- The port number for the fleet manager server -->
+      <arg name="port" value="8003"/>
```
Done! The port number for each fleet is now only accessible via the respective `config.yaml` files

xiyuoh replied:
Done, `base_url` and `port` values are now only accessible from the respective fleet `config.yaml` file

xiyuoh commented on rmf_demos/config/airport_terminal/cleanerBotA_config.yaml:
```diff
@@ -0,0 +1,69 @@
+
+# TRANSFORM CONFIG =============================================================
+# For computing transforms between Robot and RMF coordinate systems
+
```
Done

xiyuoh replied:
Done

xiyuoh commented on rmf_demos/config/airport_terminal/cleanerBotA_config.yaml:
```diff
@@ -0,0 +1,69 @@
+    loop: False
+    delivery: False
+    clean: True
+    finishing_request: "charge" # [park, charge, nothing]
```
Done

xiyuoh commented on rmf_demos/config/airport_terminal/tinyRobot_config.yaml:
```diff
@@ -0,0 +1,116 @@
+      charger:
+        waypoint: "n13"
+  # Configuration for tinyRobot_4
+  tinyRobot_4:
```
Done

xiyuoh commented on rmf_demos/launch/include/adapters/cleanerBotA_adapter.launch.xml:
```diff
@@ -5,12 +5,10 @@
+    <include file="$(find-pkg-share rmf_demos_fleet_adapter)/launch/fleet_adapter.launch.xml">
 
       <!-- The name and control type of the fleet -->
       <arg name="fleet_name" value="$(var fleet_name)"/>
```
Done! Removed `control_type` and `fleet_name` from launch files, changed the source of `fleet_name` to `config.yaml` from within the code

xiyuoh commented on rmf_demos_fleet_adapter/package.xml:
```diff
@@ -0,0 +1,19 @@
+  <name>rmf_demos_fleet_adapter</name>
+  <version>0.0.0</version>
+  <description>A fleet adapter for interfacing with robots with a fleet manager via REST API</description>
+  <maintainer email="yadunund@openrobotics.org">Yadunund</maintainer>
```
Done

xiyuoh commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,620 @@
+                                else:
+                                    self.on_lane = None  # update_off_grid()
+                                    self.on_waypoint = None
+                        # ------------------------ #
```
Done

xiyuoh commented:
@Yadunund Sorry for taking so long, thanks for the patience in adding the individual comments too. All the changes have been addressed! The tracker is still behaving weirdly during docking though, it skips ahead of the robot straight to the docking waypoints. 

https://user-images.githubusercontent.com/30265743/144295442-d0616277-051e-4057-948e-8ec43814a86f.mov

Yadunund reviewed (commented):
@xiyuoh fantastic work addressing all the previous comments! I made some small tweaks to improve performance at higher simulation RTF. Regarding the issue with the docking trajectory, this was resolved by inserting the current position of the robot at the start of the remaining docking waypoints.

Before merging, it would be great if more folks could run various scenarios for the different maps to report any issues. @aaronchongth @youliangtan do you mind giving this a go? Just launch the different worlds as per usual and try submitting a bunch of tasks.

youliangtan commented:
Hi :wave: Thanks for the awesome work! I briefly tested some of the scenarios. Some comments:
1) **fastapi dep installation**: This is relevant to both `fastapi` and `uvicorn`. Since the `rmf_demos::fleet_manager` uses fastapi, we will need to include this into the main installation by having this defined in the `package.xml`. Unfortunately, afaik `fastapi` is not available as a debian package.
2) **Task execution for deliveryRobot**: I briefly tried to run the `airport_terminal.world` and `hotel.world`. The `deliveryRobot` for both worlds are not moving well. There's a lot of hesitation in moving. I can attach a short clip if this description sounds vague to you.
3) **robot pose marker on rviz**: As compared to the previous method. the current robot pose marker (purple marker) doesn't get updated as frequently as before. This might not be severe. However previously on rviz, when the robot is moving, the marker will move smoothly without any noticeable leap.

I will continue to test more scenarios out. Will add more comments if I face any difficuitlies.

Yadunund commented:
@youliangtan thanks for the feedback
>fastapi dep installation: 

Can we use rosdep to install these modules?

>Task execution for deliveryRobot:

Noted. Will try this out. What command did you send out?

>robot pose marker on rviz:

We're using the default 1Hz frequency to publish fleet states using a wall timer inside FleetUpdateHandle https://github.com/open-rmf/rmf_ros2/blob/a88fd242009ca1f92b4ccc4f61940053f4440905/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/FleetUpdateHandle.hpp#L179-L183
I guess we could increase this frequency for smoother visaulization. Previously, the Fleet states were being published by the `robot_state_aggregator` which was triggered on receiving a RobotState message from `slotcar`. As RTF increases, slotcar will publish RobotState at a higher rate and hence the FleetState was also being publihsed at the similar rate (by virtue of the callback execution frequency). But the good news is that the 1Hz FleetState update is purely visual. Internally, the `RobotUpdateHandle::update()` method is called at a much higher rate and it scales as RTF increases.

youliangtan commented:
> Can we use rosdep to install these modules?

As from [rosdep](https://github.com/ros/rosdistro/blob/master/rosdep/python.yaml), it's not available. We will need to manually pip3 install it, which is not that ideal.

> Noted. Will try this out. What command did you send out?

you can try `ros2 run rmf_demos_tasks dispatch_loop -s kitchen -f kitchen --use_sim_time` in the `hotel.world`. Btw, `tinyrobot` works perfectly fine.

> robot pose marker on rviz

i guess a smoother visualization is preferred? =)

youliangtan commented:
IIRC, i faced the same problem with the deliverybot in clinic world too (even using the legacy fleet adapter). I guess the 2 delivery robots in the clinic world are placed too closely, thus slotcar not able to move it? 🤔

xiyuoh commented:
@youliangtan yes that's the problem! Had a short chat with Grey and he suggested parking deliveryRobot_1 further away as a temporary fix, as there will be future RMF changes addressing this issue. I've updated `clinic.building.yaml` and the config file accordingly

youliangtan commented on rmf_demos_fleet_adapter/package.xml:
```diff
@@ -0,0 +1,20 @@
+  <depend>rclpy</depend>
+  <depend>rmf_fleet_msgs</depend>
+  <depend>rmf_task_msgs</depend>
+
```
After https://github.com/ros/rosdistro/pull/31624 is merged, should add this for rosdep installation.
```
  <depend>univorn-pip</depend>
  <depend>fastapi-pip</depend>
```
or we can create a stopgap for this, by running the legacy fleet adapter by default. If the person would like to run the new fleetadapter and manager, they just need to swich the launch file, and install `pip3 install fastapi uvicorn` manually.

xiyuoh replied:
Thanks for the rosdep PR and suggestion! I've updated the launch files with the toggle for the new fleet manager and instructions in `rmf_demos_fleet_adapter`'s readme accordingly

youliangtan reviewed (changes requested):
Tested a few scenarios again. I think this definitely runs better than previously. :smile: Here I have 2 more comments:
1) fleet update freq: refer to comment
2) rosdep of fastapi: Regarding this, I have opened a pr to rosdep. I think it will take a while before getting merged.  As a stopgap, If we would prefer to merge this demo pr earlier, we can make the "old" fleet adapter as default. User can easily switch to "fleet manager mode" via roslaunch (e.g.: `enable_fleet_manager_mode:=1`). Regarding this, we just need to add some instructions to the main `README.md`.

xiyuoh commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_adapter.py:
```diff
@@ -0,0 +1,329 @@
+    fleet_handle = adapter.add_fleet(fleet_name, vehicle_traits, nav_graph)
+
+    if not fleet_config['publish_fleet_state']:
+        fleet_handle.fleet_state_publish_period(None)
```
I tried this as well, but `robot_state_update_frequency` overrides it. Perhaps `robot_state_update_frequency` is sufficient to update the individual robot's state publish frequency

youliangtan replied:
In terms of rviz visualization, a `fleet_state_visualizer` node will subscribe to `fleet_state` topic, then publishes `fleet_markers` to rviz. So, in order to increate the "fps" on rviz, the fleet state need to be published with a higher freq. Certainly, individual robot's `robot_state_update_frequency` needs to be higher than the fleet pub rate. 
I tried it on my local pc, it works by changing `fleet_state_publish_period` to a lower duration.

xiyuoh replied:
Ah thanks for explaining! I will add it in then. For some reason my rviz visualization is still able to update at a higher frequency without changing the `fleet_state_publish_period`.

youliangtan commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_manager.py:
```diff
@@ -0,0 +1,351 @@
+                    return data
+                data['data'] = self.get_robot_state(state, robot_name)
+            data['success'] = True
+            return data
```
Wanna sneak in another comment. It is preferred to specify the data model of the response msg, [example](https://fastapi.tiangolo.com/tutorial/response-model/). The data model will have the same structure as what is mentioned on the [readme](https://github.com/xiyuoh/rmf_demos/tree/feature/demo_fleet_adapter/rmf_demos_fleet_adapter#response-body-1)

This will be helpful for users to view the format of all endpoints via the swagger UI. (e.g. when running office world,  http://127.0.0.1:8001/docs). This recommendation also applies to other requests and response types of all exisitng fastapi endpoints.

xiyuoh replied:
:+1: got it and done!

youliangtan commented:
Thanks for addressing the comments. 2 minor comments from me:
- Since the instruction for running the `rmf_demos_fleet_manager` is hidden within then `rmf_demos_fleet_adapter`'s readme, it will be good if we can mention the `rmf_demos_fleet_adapter`'s README explicitly in the main README.
- I noticed when I ran a couple of scenarios on my machine, there was "motion blur" on rviz when running `LoopRequest` (doesn't happen with `Clean`). this can be seen clearly when speeding up the sim. I guess something to do with the publishing of fleetState?
![Screenshot from 2022-01-04 09-21-03](https://user-images.githubusercontent.com/32189404/147997844-945ea080-3505-4434-a2f5-3609dcd62d9d.png)

Other than that, I think the rest looks good to me! I think @Yadunund will chip in and make a final review before finally merging in.

xiyuoh commented:
@youliangtan Thanks for the comments! I've updated the main readme with a small section mentioning the `rmf_demos_fleet_adapter` under Additional Features. For the blurred marker, it happens when `robot_state_aggregator` is launched together with the fleet manager, i suppose their publishing frequency is different like you said. I moved it into the adapter launch files so that it will only be launched with the legacy adapter.

Yadunund commented:
> @youliangtan Thanks for the comments! I've updated the main readme with a small section mentioning the `rmf_demos_fleet_adapter` under Additional Features. For the blurred marker, it happens when `robot_state_aggregator` is launched together with the fleet manager, i suppose their publishing frequency is different like you said. I moved it into the adapter launch files so that it will only be launched with the legacy adapter.

We should not be running `robot_state_aggregator` along with the new fleet adapters.

Yadunund commented:
> 2. rosdep of fastapi: Regarding this, I have opened a pr to rosdep. I think it will take a while before getting merged.  As a stopgap, If we would prefer to merge this demo pr earlier, we can make the "old" fleet adapter as default. User can easily switch to "fleet manager mode" via roslaunch (e.g.: `enable_fleet_manager_mode:=1`). Regarding this, we just need to add some instructions to the main `README.md`.

@youliangtan imo supporting the old adapters goes against the motivation of implementing the new adapters and will further confuse users on how the robots are being controlled (event if its a stopgap). If the intention is avoid installation errors due to rosdep keys not yet being available, why not update the installation step [here](https://github.com/open-rmf/rmf#additional-dependencies) with instructions on installing fastapi and other dependencies? Then once the rosdep PR is merged in, we can update the instructions again.

youliangtan commented:
Fairpoint! We can surely completely remove the legacy implementation since user should always intergrate their fleet with the existing `rmf_fleet_adapter` cpp/python api. We can surely open a pr in `open-rmf/rmf` on this. 

Meanwhile, @xiyuoh would like to trouble you to remove the legacy impl of fleet adapter here. Side note, the CI [here](https://github.com/open-rmf/rmf_demos/blob/2371649188c608eb3191cb4cb829efa067202b6c/.github/workflows/build.yaml#L20) will also require an update.

xiyuoh commented:
@youliangtan done and done! I removed the portion about different fleet adapter modes from both READMEs as well

Yadunund reviewed (changes requested):
This is looking really great. I only have a few tiny nitpicks to address. Apart from that, I recommend setting `filter_waypoints` to `False`. The original intention of this function was to remove waypoints that represent rotations on the spot. But this leads to a lot discrepancy on what the robot is actually doing vs what the fleet adapter thinks it is doing. As a result, negotiations become harder to resolve. Even when set to `False`, the function will perform some touch ups to prevent robots from "backtracking" as seen [here](https://github.com/xiyuoh/rmf_demos/blob/b9a9edc7f508e1e1a9f9ebd3e9233ba70c79afe4/rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py#L548-L563) and this is what is more useful.

youliangtan commented on rmf_demos_fleet_adapter/README.md:
````diff
@@ -0,0 +1,171 @@
+source ~/rmf_ws/install/setup.bash
+ros2 launch rmf_demos_gz office.launch.xml
+```
+Then, visit http://127.0.0.1:8001/docs in your browser to interact with the endpoints.
````
I think `rmf-web`'s reporting server is using 8002 and 8003. Let's change the port number in the config. Use some less common port number like : `11001` or `7001`

xiyuoh replied:
:+1: done!

xiyuoh commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/fleet_adapter.py:
```diff
@@ -0,0 +1,330 @@
+
+    fleet_state_update_frequency = fleet_config['publish_fleet_state']
+    fleet_handle.fleet_state_publish_period(
+        datetime.timedelta(milliseconds=1000.0/fleet_state_update_frequency))
```
:+1: done!

xiyuoh commented on rmf_demos/launch/airport_terminal.launch.xml:
```diff
@@ -14,34 +14,19 @@
 
   <!-- TinyRobot fleet adapter, and robot state aggregator needed for the TinyRobot slotcar_plugin -->
```
:+1: done!

xiyuoh commented on rmf_demos_fleet_adapter/README.md:
```diff
@@ -0,0 +1,171 @@
+
+This is an implementation of the python based [fleet adapter template](https://github.com/open-rmf/fleet_adapter_template) on selected RMF demo worlds: Hotel, Office, Airport Terminal and Clinic. 
+
+The fleet adapter uses FastAPI to:
```
Got it, updated the readme. I hope the brief explanation is sufficient, didn't elaborate too much on the endpoints since that is done in the following sections

Yadunund reviewed (dismissed):
Thanks for addressing all the points. Everything looks great to me 👍🏼

aaronchongth commented on rmf_demos_maps/maps/clinic/clinic.building.yaml:
```diff
@@ -1015,7 +1015,7 @@ levels:
       - [627.408, 1497.723, 0, ""]
       - [499.965, 1254.912, 0, ""]
-      - [264.658, 1186.373, 0, L1_sub_waiting_area_1, {is_holding_point: [4, true], is_parking_spot: [4, false]}]
+      - [264.658, 1186.373, 0, L1_sub_waiting_area_1, {is_charger: [4, true], is_holding_point: [4, true], is_parking_spot: [4, true], spawn_robot_name: [1, deliveryRobot_1], spawn_robot_type: [1, DeliveryRobot]}]
```
I see a duplicate spawning of `deliveryRobot_1`, which causes the world in `rmf_demos_ign` to fail during start-up

xiyuoh replied:
Thanks for pointing it out! I removed the duplicate spawn and ignition launches properly now

luca-della-vedova reviewed (dismissed):
I tested on office Gazebo / Ignition and it looks good, I got a few deadlocks but from my understanding they are cause by slotcar's stop radius being too conservative.
I just have a small few comments on dependencies and ideas on how to reduce the diff in areas where it can be done

Yadunund commented on rmf_demos_fleet_adapter/README.md:
````diff
@@ -0,0 +1,171 @@
+
+Make sure the required dependencies are installed:
+```bash
+pip3 install fastapi uvicorn
````
I've asked @xiyuoh to get rid of the nudged dependency as we don't really need to do any coordinate transformations

xiyuoh replied:
Done!

Yadunund commented on rmf_demos_fleet_adapter/rmf_demos_fleet_adapter/RobotCommandHandle.py:
```diff
@@ -0,0 +1,618 @@
+            self.on_waypoint = start.waypoint
+
+        transient_qos = QoSProfile(
+            history=History.RMW_QOS_POLICY_HISTORY_KEEP_LAST,
+            depth=1,
+            reliability=Reliability.RMW_QOS_POLICY_RELIABILITY_RELIABLE,
+            durability=Durability.RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL)
```
I think I tried updating to the `galactic` api to get rid of these warnings but then had trouble running on `foxy`. I think we're still trying to support `foxy`?

youliangtan commented on rmf_demos_fleet_adapter/package.xml:
```diff
@@ -0,0 +1,20 @@
+  <author email="xiyu@openrobotics.org">Xi Yu Oh</author>
+  <license>Apache License 2.0</license>
+
+  <exec_depend>rmf_fleet_adapter_python</exec_depend>
```
Yea i kinda agree with this. And each of us will have different preinstalled py dependencies. One way to test this on a clean machine, or build&run the demo in a docker image, refered to the dockerfile in `open-rmf/rmf`.

xiyuoh replied:
Hello! I tried building and running this with the open-rmf/rmf docker image, it works fine 😊

Yadunund commented on rmf_demos/launch/include/hotel/hotel.rviz:
```diff
@@ -6,16 +6,18 @@ Panels:
       Expanded:
         - /Status1
         - /MarkerArray1
+        - /MarkerArray1/Status1
```
The changes to QoS settings for the flooorplan subscription are essential.

youliangtan commented:
On a side observation, when I try to dispatch the white `DeliveryRobot` by increasing the step size to speed up the sim: `gz physics -s 0.01`, the robot won't move. Other robot models (e.g. tinybot, cleaning bot) work perfectly fine with a higher step size. This deliveryrobot issue also happens in multiple worlds. Although this is not crucial, is it possible to quick fix this as this is handy for "less powerful" PC.

youliangtan commented:
@Yadunund However, this `deliveryrobot` issue doesn't occur to the previous method of running fleet adapter that's why I am raising this up. We could also address this separtely in a diff pr. Also for the contact points of `deliveryrobot`, it seems that both the center driving wheels are stable in contact with the floor. Ideally the model should fine.

Yadunund commented:
@youliangtan interesting observation. I'm not sure why the changes in this PR would cause this behavior, especially for just one of the robots but it is worth exploring further.

Yadunund commented:
@xiyuoh we also need to update the fleet_adapter.py file + configs to parse rmf-web's websocket server URI. Then we need to pass along with we [add a fleet](https://github.com/open-rmf/rmf_ros2/blob/9b4b8a8cc38b323f875a55c70f307446584d1639/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/Adapter.hpp#L104). This is an optional field so we do this only if the config has this parameter. For testing, you can pass `ws://localhost:8001` and run `rmf-web` to make sure we can get the robot states + trajectories visualized on the dashbaord

Yadunund commented:
It just occurred to me that if we pass in the `server_uri` and we don't run rmf-web's API server, the fleet adapter will continuously print out warning messages that it's unable to establish connection to the server. This might be alarming. So could we modify the implementation such that
1. We declare `server_uri` as a ROS 2 Param with default value set to an empty string. If a nonempty string is provided we pass that along to `add_fleet()`, else pass `None`.
2. Update the adapter launch file to store a `server_uri` argument with default value as an empty string. This arg is passed as a param to the fleet adapter node. Add a similar argument to the top level simulation launch file that gets passed down to the adapter launch file. 

This way users can choose to run with or without connecting to the server at run time. For example
`ros2 launch rmf_demos_gz office.launch.xml server_urI:=ws://127.0.0.1:8001` if they want to connect to the server and without the argument overwrite if they do not wish to connect.

Sorry about the extra work!

Yadunund reviewed (dismissed):
Thanks for all the iteration. Everything looks good to me. Happy to merge in if nobody else has any objections.

luca-della-vedova commented:
I think a part that I'm still slightly concerned about is handling dependencies, specifically the `fastapi` / `uvicorn` dependencies.
Since the [PR in rosdistro](https://github.com/ros/rosdistro/pull/31624) has been merged, we could add both of them as dependencies to the package.xml to make sure rosdep installs them and we don't need to have additional manual steps. However, in Ubuntu 22.04 the fastapi key resolves to the system one that is a much older version than `pip` (0.63.0 vs 0.78.0).
I gave it a quick try and the fleet adapter crashes with such an old version of `fastapi` which means we can't quite add it to the dependencies and have it managed by rosdep.
I was wondering how much work it would be to have it compatible with both versions and would this be an acceptable solution? From what I saw in the comment in the [rosdistro PR](https://github.com/ros/rosdistro/pull/31624#issuecomment-1022652210), packages can't be released if they depend on a `pip` package so that might be a bit of a deal breaker

luca-della-vedova commented:
Upon further investigation, this seems to be due to an issue with the Ubuntu repositories which is not really solvable right now.

I reported the issue and a possible solution in the [launchpad for fastapi](https://bugs.launchpad.net/ubuntu/+source/fastapi/+bug/1970557?comments=all), hopefully it can be fixed there, not much we can do for now.

Yadunund reviewed (approved):
Thanks for making the changes and thanks to everyone else for helping review.
